### DISCUSSION TITLE
Drop Shadow

### DISCUSSION THREAD

**Discussion by GhishNchips** [YaLTeR/niri#851] [Ideas]
Hi!

It would be nice if we could have drop shadows for windows! With settings similar to https://github.com/yshui/picom or such!

**Comment by YaLTeR** [comment]
Maybe yeah. This may become more important with floating windows.

**Reply by phuhl** [reply]
Yup, just noticing this while testing floating windows. I ran without border until now and that worked very well for just tiling, but same colored windows on top of each other start to be an issue without a border or shadow.

Personally, I'd prefer a shadow over borders, so seconding this feature request :)

**Reply by YaLTeR** [reply]
Made an issue: https://github.com/YaLTeR/niri/issues/927

**Comment by Yaroslav-95** [comment]
Thanks a lot for the new shadow feature! I have a question, though, is there a way to configure it for layer surfaces (e.g. notifications, menus) or are the shadows just for windows?

**Reply by YaLTeR** [reply]
Only toplevels draw shadows right now. What do you have in mind?

**Reply by Yaroslav-95** [reply]
I was mostly thinking about notification pop-ups, more specifically mako's. Right now they aren't getting any shadows drawn around them, and I was wondering if that was intentional or if it required further configuration.

**Reply by YaLTeR** [reply]
I'm not sure mako can work properly with compositor-side decorations (including shadows) due to how it draws its layer surfaces and how layer-shell protocol doesn't have a way to indicate surface geometry.

**Reply by Yaroslav-95** [reply]
I am not very well versed in the Wayland protocol yet, but I can say that in swayfx shadows can be drawn around mako's notifications, and other layer-shell surfaces. They do have a separate configuration option for layers, in which you can individually enable/disable effects (i.e. including shadows) for a layer namespace (e.g. "notifications", "rofi", etc.)
Also, if I might say so, layer-shell surfaces do have to tell the compositor their geometry, otherwise how would the compositor know how to draw them? I took a peek at mako's code, and they, for example, use the [zwlr_layer_surface_v1::set_size](https://wayland.app/protocols/wlr-layer-shell-unstable-v1#zwlr_layer_surface_v1:request:set_size) request, which tells the compositor mako's surface(s) geometr(y/ies). Or do you mean something else by geometry like the (non-rectangular?) shape of the surface?

**Reply by YaLTeR** [reply]
Yes, the toplevel geometry tells the compositor the window visual rectangle as it appears to the user, i.e. excluding any client-side shadows or padding or whatever: https://wayland.app/protocols/xdg-shell#xdg_surface:request:set_window_geometry

There's no such thing for layer-shell. And for example mako with gaps will draw its surface all the way to the screen edge, including gaps. There's no way for it to tell the compositor that these are gaps and should be excluded from the shadow.

If swayfx doesn't have a problem with this, I wonder how they do it

**Reply by Yaroslav-95** [reply]
I see what you mean. I modified my mako config to add a margin between notifications and it, indeed broke, not only the shadows, but the blur as well, lol.
That said, maybe it would still be feasible to add an option to selectively enable shadows for different layer namespaces just like swayfx does? Just have the shadows be disabled for all layer namespaces by default and let the user decide if they want the shadows to be drawn for specific components. What do you think?

**Reply by YaLTeR** [reply]
That should work I guess, yeah

**Reply by YaLTeR** [reply]
Implemented in acd4cb51aa0e013bfec14444ee48a01c60ebaf8a

**Reply by Yaroslav-95** [reply]
Wow, that was really fast! Thanks a lot!

P.S: я вижу у вас есть Github sponsors, а есть ли способ пожертвовать без зарубежной карты?

**Reply by YaLTeR** [reply]
Могу наверное прислать инфу куда-нибудь в личку)

**Reply by Yaroslav-95** [reply]
Отправил письмо на емейл.
